sanic_routing/tree.py

```diff
@@ -258,111 +333,79 @@ def _inject_requirements(self, location, indent):
+                ),
             ]
         )
 
-    def _has_nested_path(self, node, recursive=True):
-        if node.group and (
-            (node.group.labels and "path" in node.group.labels)
-            or (node.group.pattern and r"/" in node.group.pattern)
-        ):
-            return True
-        if recursive and node.children:
-            for child in node.children.values():
-                if self._has_nested_path(child):
-                    return True
-        return False
-
-    def _has_nested_regex(self, node, recursive=True):
-        if node.group and node.group.regex:
-            return True
-        if recursive and node.children:
-            for child in node.children.values():
-                if self._has_nested_regex(child):
-                    return True
-        return False
-
-    def _sorting(self, item) -> t.Tuple[bool, int, str, bool, int]:
+    def _sorting(self, item) -> t.Tuple[bool, bool, int, int, int, bool, str]:
+        """
+        Primarily use to sort nodes to determine the order of the nested tree
+        """
         key, child = item
         type_ = 0
         if child.dynamic:
-            key = key[1:-1]
-            if ":" in key:
-                key, param_type = key.split(":", 1)
-                try:
-                    type_ = list(self.router.regex_types.keys()).index(
-                        param_type
-                    )
-                except ValueError:
-                    type_ = len(list(self.router.regex_types.keys()))
+            type_ = child.param.priority
+
         return (
+            bool(child.groups),
             child.dynamic,
+            type_ * -1,
+            child.depth * -1,
             len(child._children),
+            not bool(
+                child.groups and any(group.regex for group in child.groups)
+            ),
             key,
-            bool(child.group and child.group.regex),
-            type_ * -1,
         )
 
+    def _group_sorting(self, item) -> t.Tuple[int, ...]:
+        """
+        When multiple RouteGroups terminate on the same node, we want to
+        evaluate them based upon the priority of the param matching types
+        """
+
+        def get_type(segment):
+            type_ = 0
+            if segment.startswith("<"):
+                key = segment[1:-1]
+                if ":" in key:
+                    key, param_type = key.split(":", 1)
+                    try:
+                        type_ = list(self.router.regex_types.keys()).index(
+                            param_type
+                        )
+                    except ValueError:
+                        type_ = len(list(self.router.regex_types.keys()))
+            return type_ * -1
+
+        segments = tuple(map(get_type, item.parts))
+        return segments
+
+    @property
+    def depth(self):
+        if not self._children:
+            return self.level
+        return max(child.depth for child in self._children.values())
```

Comment:
Is this off by one, since it is directly using child's depth?

Reply:
Nope. If it has children, it should get the depth of the deepest child and use that.

Here is an illustration of the intent:

```
Node (1)
Node (4)
    Node (3)
        Node (3)
    Node (4)
        Node (4)
            Node (4)
        Node (3)
```